### PR TITLE
Updated MLPF producer with ONNX

### PR DESCRIPTION
**Pull request by jpata**
#### PR description:

Following the presentation at the [PPD general meeting](https://indico.cern.ch/event/1094349/contributions/4602823/attachments/2344166/3996970/cms_mlpf_acat2021.pdf), at [ACAT2021](https://indico.cern.ch/event/855454/contributions/4597457/), and as outlined in the [PPD workshop](https://indico.cern.ch/event/1103515/#17-mlpf-status-plans-and-ideas), we are updating the MLPF integration in CMSSW to facilitate further development and scrutiny.

Note that MLPF is **off by default** and thus no changes are expected in any of the standard workflows.

This PR mainly updates the ML model and switches the inference to ONNX from tensorflow. MLPF-specific event content is removed, as now MLPF produces PFCandidates instead of (rather than in parallel to) PFAlgo, when enabled.

Here are the igprof results:
- memory: https://jpata.web.cern.ch/jpata/cgi-bin/igprof-navigator/mlpf_v2/ig_mem
- CPU: https://jpata.web.cern.ch/jpata/cgi-bin/igprof-navigator/mlpf_v2/ig_cpu

#### PR validation:

For physics validation, please see the slides linked above. The integration can be tested in the workflows 11843.13 and 11834.13.

### CODE REVIEW

**Comment by jpata** [comment]
test parameters:
  - pull_requests = https://github.com/cms-data/RecoParticleFlow-PFProducer/pull/3
  - workflows = 11834.13, 11843.13

**Comment by jpata** [comment]
@cmsbuild please test

**Comment by jpata** [comment]
test parameters:
  - pull_requests = https://github.com/cms-data/RecoParticleFlow-PFProducer/pull/3
  - workflows = 11834.13, 11843.13
  - relvals_opt = --what upgrade,standard,highstats,pileup,generator,extendedgen,production,ged,machine,premix

**Comment by jpata** [comment]
@cmsbuild please test

(the new workflows are in the upgrade matrix, I'm trying to get the bot to run them based on https://github.com/cms-sw/cmssw/pull/36748#issuecomment-1017755649. let me know if there is a better way!)

**Comment by srimanob** [comment]
The error is on era which is introduced and just merged in https://github.com/cms-sw/cmssw/pull/36753

**Comment by jpata** [comment]
> The error is on era which is introduced and just merged in #36753

Thanks for confirming. I see that the workflows 11834.13 and 11843.13 ran successfully, which was the intended test.

**Comment by clacaputo** [comment]
@cmsbuild please test

**Comment by jpata** [comment]
@cmsbuild please test

**Comment by jpata** [comment]
adding @cms-sw/pf-l2 @laurenhay here, just to make sure everyone is informed.

**Comment by jfernan2** [comment]
+1

**Comment by clacaputo** [comment]
+reconstruction

- no changes are expected since MLPF is off by default
- WFs 11834.13 and 11843.13 run without problems

**Comment by srimanob** [comment]
+Upgrade

From the code related to Upgrade, updating the MLPF workflow to allow QCD sample is fine.

**Comment by kskovpen** [comment]
+pdmv

File: RecoParticleFlow/PFProducer/plugins/MLPFProducer.cc [hunk 1]
```diff
@@ -4,135 +4,175 @@
...
   assert(tensor_size <= NUM_MAX_ELEMENTS_BATCH);
+  assert(tensor_size % LSH_BIN_SIZE == 0);
+#endif
 
-  //Create the input tensor
-  tensorflow::TensorShape shape({BATCH_SIZE, tensor_size, NUM_ELEMENT_FEATURES});
-  tensorflow::Tensor input(tensorflow::DT_FLOAT, shape);
-  input.flat<float>().setZero();
+#ifdef MLPF_DEBUG
+  std::cout << "tensor_size=" << tensor_size << std::endl;
```
Comment (perrotta) [inline RecoParticleFlow/PFProducer/plugins/MLPFProducer.cc:62]: Is this needed, given the previous `assert`'s?
If really needed to debug, then maybe better to have this `cout` before [L55](https://github.com/cms-sw/cmssw/pull/36841/files#diff-2c58f31fd9c90f205173f66905a6a529b38cefb33850b85340f03e795e44c1caR55)

File: RecoParticleFlow/PFProducer/src/MLPFModel.cc [hunk 2]
```diff
@@ -174,10 +234,28 @@ namespace reco::mlpf {
 
     math::PtEtaPhiELorentzVectorD p4(pred_pt, pred_eta, pred_phi, pred_e);
 
-    reco::PFCandidate cand(
-        0, math::XYZTLorentzVector(p4.X(), p4.Y(), p4.Z(), p4.E()), reco::PFCandidate::ParticleType(0));
-    cand.setPdgId(pred_pid);
-    cand.setCharge(charge);
+    reco::PFCandidate::ParticleType particleType(reco::PFCandidate::X);
+    if (pred_pid == 211)
```
Comment (perrotta) [inline RecoParticleFlow/PFProducer/src/MLPFModel.cc:238]: It looks like this `pred_pid` is unsigned (here and everywhere else in this code), i.e. it does not consider the charge: can you confirm, just for check<ì

Comment (jpata) [reply RecoParticleFlow/PFProducer/src/MLPFModel.cc:238]: Yes, by construction, the underlying model reconstructs the absolute value of the PID separately from the charge.

So far, we didn't present detailed results about the charge prediction (so `//cand.setCharge(charge);` downstream), but it should not be a major problem, as should be driven by the track information.

**Comment by perrotta** [comment]
+1
- It goes together with https://github.com/cms-data/RecoParticleFlow-PFProducer/pull/3

**Pull request merged**